lib/turbolinks/assertions.rb

```diff
@@ -0,0 +1,39 @@
+module Turbolinks
+  module Assertions
+    TURBOLINKS_VISIT = /Turbolinks\.visit\("([^"]+)", {"action":"([^"]+)"}\)/
+
+    def assert_redirected_to(options = {}, message = nil)
+      if turbolinks_request?
+        assert_turbolinks_visited(options, message)
+      else
+        super
+      end
+    end
+
+    def assert_turbolinks_visited(options = {}, message = nil)
+      assert_response(:ok, message)
+      assert_equal("text/javascript", response.content_type)
+
+      visit_location, visit_action = turbolinks_visit_location_and_action
+
+      redirect_is       = normalize_argument_to_redirection(visit_location)
+      redirect_expected = normalize_argument_to_redirection(options)
+
+      message ||= "Expected response to be a Turbolinks visit to <#{redirect_expected}> but was a visit to <#{redirect_is}>"
+      assert_operator redirect_expected, :===, redirect_is, message
+    end
+
+    def turbolinks_request?
+      # Technically we'd check that Turbolinks-Referrer request header is
+      # also set, but that'd require us to pass the header from post/patch/etc
+      # test methods as well.
+      request.xhr? && !request.get?
```

Comment:
Can't check `request.xhr?` after the request is processed, unfortunately, because we [unset X-Requested-With afterward](https://github.com/rails/rails/commit/4b72a4b12fbb33d32ae3f9b9d004ca5888d42d58) so it doesn't leak into subsequent requests.

Reply:
Switching to detecting the response Content-Type 😑